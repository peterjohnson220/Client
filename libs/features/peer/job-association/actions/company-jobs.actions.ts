import { Action } from '@ngrx/store';
import { GridDataResult } from '@progress/kendo-angular-grid';

export const LOAD_COMPANY_JOBS  = '[Peer Job Association Modal/Company Jobs Grid] Load Company Jobs';
export const LOAD_COMPANY_JOBS_SUCCESS  = '[Peer Job Association Modal/Company Jobs Grid] Load Company Jobs Success';
export const LOAD_COMPANY_JOBS_ERROR  = '[Peer Job Association Modal/Company Jobs Grid] Load Company Jobs Error';
export const SEARCH_TERM_UPDATED = '[Peer Job Association Modal/Company Jobs Grid] Search Term Updated';

export class LoadCompanyJobs implements Action {
  readonly type = LOAD_COMPANY_JOBS;
}
export class LoadCompanyJobsSuccess implements Action {
  readonly type = LOAD_COMPANY_JOBS_SUCCESS;
  constructor(public payload: GridDataResult) {}
}

export class LoadCompanyJobsError implements Action {
  readonly type = LOAD_COMPANY_JOBS_ERROR;
}

export class SearchTermUpdated implements Action {
  readonly type = SEARCH_TERM_UPDATED;
  constructor(public payload: string) {}
}

export type Actions
  = LoadCompanyJobs
  | LoadCompanyJobsSuccess
  | LoadCompanyJobsError
  | SearchTermUpdated;
