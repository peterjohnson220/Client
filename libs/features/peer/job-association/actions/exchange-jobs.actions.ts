import { Action } from '@ngrx/store';
import { GridDataResult } from '@progress/kendo-angular-grid';

export const LOAD_EXCHANGE_JOBS  = '[Peer Job Association Modal/Exchange Jobs] Load Exchange Jobs';
export const LOAD_EXCHANGE_JOBS_SUCCESS  = '[Peer Job Association Modal/Exchange Jobs] Load Exchange Jobs Success';
export const LOAD_EXCHANGE_JOBS_ERROR  = '[Peer Job Association Modal/Exchange Jobs] Load Exchange Jobs Error';

export class LoadExchangeJobs implements Action {
  readonly type = LOAD_EXCHANGE_JOBS;
}

export class LoadExchangeJobsSuccess implements Action {
  readonly type = LOAD_EXCHANGE_JOBS_SUCCESS;
  constructor(public payload: GridDataResult) {}
}

export class LoadExchangeJobsError implements Action {
  readonly type = LOAD_EXCHANGE_JOBS_ERROR;
}

export type Actions
  = LoadExchangeJobs
  | LoadExchangeJobsSuccess
  | LoadExchangeJobsError;
