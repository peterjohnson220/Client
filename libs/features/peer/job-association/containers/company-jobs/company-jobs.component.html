<h2 class="text-center mb-3">Company Jobs</h2>
<div class="filters position-relative d-flex mt-2 mb-2">
  <fa-icon icon="search" class="search-icon position-absolute"></fa-icon>
  <pf-input-debounce
    [id]="'company-jobs-search-box'"
    [delay]="400"
    [placeholderText]="'Search job title or job code'"
    (valueChanged)="handleSearchBoxValueChanged($event)">
  </pf-input-debounce>
</div>
<div>
  <label class="text-uppercase" data-qa-id="lbl-company-jobs-total">
    company jobs: {{ totalCompanyJobsGridItems$ | async }} total | selected: {{selectedCompanyJobIds.length}}
  </label>
</div>
<div class="grid-container position-relative">
  <pf-async-container
    class="h-100"
    [loading]="loading$ | async"
    [loadingError]="loadingError$ | async"
    [loadingErrorMessage]="'Error loading company jobs'"
    (reload)="reload()"
  >
  <div class="k-tooltip-container"
       kendoTooltip
       showOn="none"
       [tooltipTemplate]="template"
       filter=".k-grid td"
       (mouseover)="showGridTooltip($event)">
    <pf-grid-detail-panel
      [isExpanded$]="isDetailPanelExpanded$"
      [checkForJdmDescription]="true"
      [jdmDescriptionIds]="jdmDescriptionIds$ | async"
      [jdmDescriptionLoading]="jdmDescriptionLoading$ | async"
      [jdmDescriptionLoadingError]="jdmDescriptionLoadingError$ | async"
      [jobId]="selectedCompanyJobInDetailPanel.CompanyJobId"
      [jobTitle]="selectedCompanyJobInDetailPanel.JobTitle"
      [jobCode]="selectedCompanyJobInDetailPanel.JobCode"
      [jobDescription]="selectedCompanyJobInDetailPanel.JobDescription"
      [jobFamily]="selectedCompanyJobInDetailPanel.JobFamily"
      (closeClick)="handleCloseDetailPanel()"
      (viewJdmDescriptionClick)="handleViewJdmDescriptionClick()"
    ></pf-grid-detail-panel>
    <kendo-grid
                [data]="companyJobsGridItemsData$ | async"
                [style.height.%]="100"
                [pageSize]="(gridState$ | async)?.take"
                [skip]="(gridState$ | async)?.skip"
                [sort]="(gridState$ | async)?.sort"
                [filter]="(gridState$ | async)?.filter"
                [selectable]="{
                  enabled: true,
                  checkboxOnly: true
                }"
                [sortable]="true"
                [resizable]="true"
                [pageable]="{
                  buttonCount: 5,
                  info: true
                 }"
                (dataStateChange)="handleDataStateChange($event)"
                [kendoGridSelectBy]="'CompanyJobId'"
                [rowSelected]="isRowSelected"
                (selectionChange)="handleSelectionChange($event)">
      <kendo-grid-column field="AssociationStatus" title="" [sortable]="false" [width]="11" class="p-0 m-0">
        <ng-template kendoGridCellTemplate let-dataItem>
          <span 
            class="association-status" 
            [ngClass]="(dataItem.IsAssociated || isPendingAssociation(dataItem.CompanyJobId)) ? 'associated' : 'not-associated'"
            [title]="(dataItem.IsAssociated || isPendingAssociation(dataItem.CompanyJobId)) ? 'Associated' : 'Not Associated'"
          >
          </span>
        </ng-template>
      </kendo-grid-column>
      <kendo-grid-checkbox-column showSelectAll="true" [resizable]="false" [width]="35" class="no-ellipsis">
        <ng-template kendoGridHeaderTemplate>
          <input class="k-checkbox" id="selectAllCheckboxId" kendoGridSelectAllCheckbox
                 [state]="getSelectAllState()"
                 (click)="handleSelectAllClick()">
          <label class="k-checkbox-label" for="selectAllCheckboxId"></label>
        </ng-template>
      </kendo-grid-checkbox-column>
      <kendo-grid-column field="CompanyJobId" title="Id" hidden="hidden"></kendo-grid-column>
      <kendo-grid-column field="JobTitle" title="Job Title">
        <ng-template kendoGridCellTemplate let-dataItem>
          <a href="javascript:void(0)" 
            (click)="handleCompanyJobTitleOrCodeClick(dataItem)" 
            class="grid-link company-job-title"
            [class.selected]="dataItem.CompanyJobId === selectedCompanyJobInDetailPanel.CompanyJobId && (isDetailPanelExpanded$ | async)">
            <div class="show-tooltip" [innerHTML]="dataItem.JobTitle | highlightText: searchTerm"></div>
          </a>
        </ng-template>
      </kendo-grid-column>
      <kendo-grid-column field="JobCode" title="Job Code">
        <ng-template kendoGridCellTemplate let-dataItem>
          <a href="javascript:void(0)" 
            (click)="handleCompanyJobTitleOrCodeClick(dataItem)" 
            class="grid-link company-job-code"
            [class.selected]="dataItem.CompanyJobId === selectedCompanyJobInDetailPanel.CompanyJobId && (isDetailPanelExpanded$ | async)">
            <div [innerHTML]="dataItem.JobCode | highlightText: searchTerm" class="show-tooltip"></div>
          </a>
        </ng-template>
      </kendo-grid-column>
      <kendo-grid-column field="JobFamily" title="Job Family" [sortable]="false">
        <ng-template kendoGridCellTemplate let-dataItem>
          <div *ngIf="!dataItem.JobFamily; else hasFamily" class="show-tooltip">No Family</div>
          <ng-template #hasFamily>
            <div class="show-tooltip">{{ dataItem.JobFamily }}</div>
          </ng-template>
        </ng-template>
      </kendo-grid-column>
      <ng-template kendoGridNoRecordsTemplate>
        {{ badRequestError.length > 0 ? badRequestError + " Please narrow down the list by using the filters above." : "No records based on your search." }}
        <a href="javascript:void(0)" (click)="reload()">
          <fa-icon icon="redo-alt"></fa-icon>Reset now
        </a>
      </ng-template>
    </kendo-grid>
  </div>
  </pf-async-container>
</div>
<ng-template #template let-anchor>
  <span>{{ anchor.nativeElement.innerText }}</span>
</ng-template>
