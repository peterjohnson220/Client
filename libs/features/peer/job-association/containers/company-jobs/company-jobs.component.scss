:host ::ng-deep .k-grid .k-grid-content td {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
.k-tooltip-container {
  height: 100%;
}
