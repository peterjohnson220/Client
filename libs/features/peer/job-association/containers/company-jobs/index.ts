export * from './company-jobs.component';
