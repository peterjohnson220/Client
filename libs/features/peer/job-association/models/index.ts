export * from './company-job.model';
export * from './exchange-job.model';
