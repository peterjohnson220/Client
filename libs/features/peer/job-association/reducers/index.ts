import { createSelector, createFeatureSelector } from '@ngrx/store';

import { IFeatureGridState } from 'libs/core/reducers/grid.reducer';
import * as fromRoot from 'libs/state/state';
import * as fromGridReducer from 'libs/core/reducers/grid.reducer';
import { GenericMenuItem } from 'libs/models/common';

// Import feature reducers
import * as fromCompanyJobsReducer from './company-jobs.reducer';
import * as fromExchangeJobsReducer from './exchange-jobs.reducer';
import * as fromJobAssociationModalReducer from './job-association-modal.reducer';

// Feature area state
export interface JobAssociationFeatureState {
  companyJobs: IFeatureGridState<fromCompanyJobsReducer.State>;
  exchangeJobs: IFeatureGridState<fromExchangeJobsReducer.State>;
  jobAssociationModal: fromJobAssociationModalReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  feature_job_association: JobAssociationFeatureState;
}

// Feature area reducers
export const reducers = {
  companyJobs: fromCompanyJobsReducer.reducer,
  exchangeJobs: fromExchangeJobsReducer.reducer,
  jobAssociationModal: fromJobAssociationModalReducer.reducer
};

// Select Feature Area
export const selectFeatureAreaState =
  createFeatureSelector<JobAssociationFeatureState>('feature_job_association');

// Feature Selectors
export const selectExchangeJobsState = createSelector(
  selectFeatureAreaState,
  (state: JobAssociationFeatureState) => state.exchangeJobs);

export const selectCompanyJobsState = createSelector(
  selectFeatureAreaState,
  (state: JobAssociationFeatureState) => state.companyJobs);

export const selectJobAssociationModalState = createSelector(
  selectFeatureAreaState,
  (state: JobAssociationFeatureState) => state.jobAssociationModal);

// Job Association Modal Selectors

export const getJobAssociationModalSaving = createSelector(
  selectJobAssociationModalState,
  fromJobAssociationModalReducer.getSaving);

export const getJobAssociationModalSavingError = createSelector(
  selectJobAssociationModalState,
  fromJobAssociationModalReducer.getSavingError);

// Exchange Jobs Selectors
export const getExchangeJobsGrid = createSelector(
  selectExchangeJobsState,
  (state: IFeatureGridState<fromExchangeJobsReducer.State>) => state.grid);

export const getExchangeJobsGridState = createSelector(
  getExchangeJobsGrid,
  fromGridReducer.getGridState);

export const getExchangeJobsFeature = createSelector(
  selectExchangeJobsState,
  (state: IFeatureGridState<fromExchangeJobsReducer.State>) => state.feature);

export const getExchangeJobsLoading = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.loading
);

export const getExchangeJobsLoadingError = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.loadingError
);

export const getExchangeJobsSearchTerm = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.searchTerm
);

export const getExchangeJobsSelectedExchangeJob = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.selectedExchangeJob
);

export const getExchangeJobsIsDetailPanelExpanded = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.isDetailPanelExpanded
);

export const {
  selectAll: getExchangeJobsList
} = fromExchangeJobsReducer.adapter.getSelectors(getExchangeJobsFeature);

export const getExchangeJobsTotal = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getTotal
);

export const getExchangeJobsData = createSelector(
  getExchangeJobsList,
  getExchangeJobsTotal,
  (data, total) => ({ data, total })
);

export const getExchangeJobsResultsCount = createSelector(
  getExchangeJobsList,
  (data) => data.length
);

// Exchange Jobs Selectors, job family filter
export const getExchangeJobsFamilyFilterLoading = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterLoading
);

export const getExchangeJobsFamilyFilterLoadingSuccess = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterLoadingSuccess
);

export const getExchangeJobsFamilyFilterLoadingError = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterLoadingError
);

export const getExchangeJobFamilyFilterIsExpanded = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterIsExpanded
);

export const getExchangeJobFamilyFilterOptions = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterOptions
);

export const getExchangeJobFamilyFilterSelectedOptionNames = createSelector(
  getExchangeJobsFeature,
  (state) => state.jobFamilyOptions.filter(o => o.IsSelected).map(o => o.DisplayName)
);

// Company Jobs Selectors
export const getCompanyJobsGrid = createSelector(
  selectCompanyJobsState,
  (state: IFeatureGridState<fromCompanyJobsReducer.State>) => state.grid);

export const getCompanyJobsGridState = createSelector(
  getCompanyJobsGrid,
  (fromGridReducer.getGridState)
);

export const getCompanyJobsFeature = createSelector(
  selectCompanyJobsState,
  (state: IFeatureGridState<fromCompanyJobsReducer.State>) => state.feature);

export const {
  selectAll: getCompanyJobsList
} = fromCompanyJobsReducer.adapter.getSelectors(getCompanyJobsFeature);

export const getCompanyJobsTotal = createSelector(
  getCompanyJobsFeature,
  fromCompanyJobsReducer.getTotal);

export const getCompanyJobsData = createSelector(
  getCompanyJobsList,
  getCompanyJobsTotal,
  (data, total) => ({ data, total})
);

export const getCompanyJobsLoading = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.loading
);

export const getCompanyJobsLoadingError = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.loadingError
);

export const getCompanyJobsSearchTerm = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.searchTerm
);

// detail panel
export const getCompanyJobsSelectedCompanyJobInDetailPanel = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.selectedCompanyJobInDetailPanel
);

export const getCompanyJobsIsDetailPanelExpanded = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.isDetailPanelExpanded
);

// associations
export const getSelectedCompanyJobs = createSelector(
    getCompanyJobsFeature,
    fromCompanyJobsReducer.getSelectedCompanyJobs);

export const getExchangeJobAssociations = createSelector(
    getExchangeJobsFeature,
    (feature) => feature.ExchangeJobAssociations
);
