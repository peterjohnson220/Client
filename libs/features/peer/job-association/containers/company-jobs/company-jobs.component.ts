// angular core
import { Component, OnDestroy, OnInit, ViewChild } from '@angular/core';

// 3rd party
import { select, Store } from '@ngrx/store';
import { Observable, Subscription } from 'rxjs';
import { TooltipDirective } from '@progress/kendo-angular-tooltip';

import { DataStateChangeEvent, GridComponent, GridDataResult, RowArgs, SelectAllCheckboxState, PageChangeEvent
} from '@progress/kendo-angular-grid';
import { State } from '@progress/kendo-data-query';

// Payfactors
import { GridTypeEnum } from '../../../../../models/common';
import { InputDebounceComponent } from '../../../../../forms/components/input-debounce';
import * as fromJobAssociationReducers from '../../reducers';
import * as companyJobsActions from '../../actions/company-jobs.actions';
import * as fromGridActions from 'libs/core/actions/grid.actions';
import { CompanyJob, ExchangeJobAssociation } from '../../models';

@Component({
  selector: 'pf-peer-job-association-company-jobs',
  templateUrl: './company-jobs.component.html',
  styleUrls: ['./company-jobs.component.scss']
})
export class CompanyJobsComponent implements OnInit, OnDestroy {
  @ViewChild(TooltipDirective) public tooltipDir: TooltipDirective;
  @ViewChild(InputDebounceComponent) public companyJobSearchComponent: InputDebounceComponent;
  @ViewChild(GridComponent) public grid: GridComponent;

  // Observables
  companyJobsGridItemsData$: Observable<GridDataResult>;
  gridState$: Observable<State>;
  totalCompanyJobsGridItems$: Observable<number>;
  loading$: Observable<boolean>;
  loadingError$: Observable<boolean>;

  // Subscriptions
  companyJobsGridItemsDataSubscription: Subscription;
  exchangeJobAssociationsSubscription: Subscription;
  selectedCompanyJobsSubscription: Subscription;
  searchTermSubscription: Subscription;

  // Properties
  companyJobGridDataResult: GridDataResult;
  exchangeJobAssociations: ExchangeJobAssociation[];
  maxSelectionThreshold: number;
  selectedCompanyJobIds: CompanyJob[];
  searchTerm;

  constructor(private store: Store<fromJobAssociationReducers.State>) {}

  ngOnInit() {
    this.maxSelectionThreshold = 10;

    this.companyJobsGridItemsData$ = this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsData));
    this.totalCompanyJobsGridItems$ = this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsTotal));
    this.gridState$ = this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsGridState));
    this.loading$ = this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsLoading));
    this.loadingError$ = this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsLoadingError));

    this.selectedCompanyJobsSubscription = this.store.pipe(select(fromJobAssociationReducers.getSelectedCompanyJobs))
      .subscribe((selectedCompanyJobs) => this.selectedCompanyJobIds = selectedCompanyJobs);

    this.companyJobsGridItemsDataSubscription =
        this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsData)).subscribe(
            (gridDataResult) => this.companyJobGridDataResult = gridDataResult);

    this.exchangeJobAssociationsSubscription =
      this.store.pipe(select(fromJobAssociationReducers.getExchangeJobAssociations)).subscribe(
            (exchangeJobAssociations) => this.exchangeJobAssociations = exchangeJobAssociations);

    this.searchTermSubscription =
      this.store.pipe(select(fromJobAssociationReducers.getCompanyJobsSearchTerm)).subscribe(
        (searchTerm) => this.searchTerm = searchTerm);

    this.store.dispatch(new companyJobsActions.LoadCompanyJobs());
  }

  ngOnDestroy() {
    this.companyJobsGridItemsDataSubscription.unsubscribe();
    this.exchangeJobAssociationsSubscription.unsubscribe();
    this.searchTermSubscription.unsubscribe();
    this.selectedCompanyJobsSubscription.unsubscribe();
  }

  reload(resetSearchTerm = false): void {
    // if this is invoked from an empty search results grid reset the term, otherwise keep the term as is and reload
    if (resetSearchTerm) {
      this.companyJobSearchComponent.clearValue();
    } else {
      this.store.dispatch(new companyJobsActions.LoadCompanyJobs());
    }
  }

  showTooltip(e: any, id: string): void {
    if (e.target.offsetWidth < e.target.scrollWidth && e.target.id.includes(id)) {
      this.tooltipDir.toggle(e.target);
    } else {
      this.tooltipDir.hide();
    }
  }

  handleDataStateChange(state: DataStateChangeEvent): void {
    this.store.dispatch(new fromGridActions.UpdateGrid(GridTypeEnum.JobAssociationModalCompanyJobs, state));
    this.store.dispatch(new companyJobsActions.LoadCompanyJobs());
  }

  handleSearchChanged(searchTerm: string) {
    this.store.dispatch(new companyJobsActions.SearchTermUpdated(searchTerm));
    if (!searchTerm  || searchTerm.length > 1) {
      this.store.dispatch(new fromGridActions.PageChange(GridTypeEnum.JobAssociationModalCompanyJobs, { skip: 0 } as PageChangeEvent));
      this.store.dispatch(new companyJobsActions.LoadCompanyJobs());
    }
  }

  handleDetailExpand(event: any): void {
    // determine how many results we have in the grid
    const gridData = this.grid.data as any;
    const totalRows = gridData.data.length;

    // collapse all rows that are not the newly expanded row so we only have one detail open at a time
    for (let i = 0; i < totalRows; i++) {
      if (i !== event.index) {
        this.grid.collapseRow(i);
      }
    }
  }

  handleSelectAllClick(): void {
    let selectedCompanyJobs: CompanyJob[] = [];
    if (this.getSelectAllState() === 'unchecked') {
      selectedCompanyJobs = this.selectCompanyJobIdsUptoThreshold();
    }
    this.store.dispatch(new companyJobsActions.SelectCompanyJobs(selectedCompanyJobs));
  }

  selectCompanyJobIdsUptoThreshold(): CompanyJob[] {
    return this.companyJobGridDataResult.data.slice(0, this.maxSelectionThreshold);
  }

  getSelectAllState(): SelectAllCheckboxState {
    if (!this.thereAreSelectedRows()) {
      return 'unchecked';
    } else if (this.hasMetMaxSelectableRowsThreshold() || this.resultsEqualSelected()) {
      return 'checked';
    } else {
      return 'indeterminate';
    }
  }

  handleSelectionChange(event: any): void {
    let selectedCompanyJobs = [...this.selectedCompanyJobIds];
    if (event.selectedRows.length > 0 && !this.hasMetMaxSelectableRowsThreshold()) {
        event.selectedRows.forEach((row) => {
          selectedCompanyJobs.push(row.dataItem);
      });
    }

    if (event.deselectedRows.length > 0) {
      event.deselectedRows.forEach((row) => {
        selectedCompanyJobs = selectedCompanyJobs
                .filter((companyJob) => companyJob.CompanyJobId !== row.dataItem.CompanyJobId);
      });
    }

    this.store.dispatch(new companyJobsActions.SelectCompanyJobs(selectedCompanyJobs));
  }

  isPendingAssociation(companyJobId: number): boolean {
    let isAssociated = false;
    this.exchangeJobAssociations.forEach((eja) => {
      if (eja.CompanyJobs.map(cj => cj.CompanyJobId).sort().indexOf(companyJobId) >= 0) {
        isAssociated = true;
        return;
      }
    });
    return isAssociated;
  }

  isRowSelected = (e: RowArgs) => this.isSelectedCompanyJob(e.dataItem.CompanyJobId);

  isSelectedCompanyJob(companyJobId: number): boolean {
    return this.selectedCompanyJobIds.map(scj => scj.CompanyJobId).indexOf(companyJobId) >= 0;
  }

  hasMetMaxSelectableRowsThreshold(): boolean {
    if (!this.selectedCompanyJobIds) {
      return false;
    }
    return (this.selectedCompanyJobIds.length > 0 && this.selectedCompanyJobIds.length === this.maxSelectionThreshold);
  }

  thereAreSelectedRows(): boolean {
    return this.selectedCompanyJobIds.length > 0;
  }

  resultsEqualSelected(): boolean {
    return (this.selectedCompanyJobIds.length > 0 &&
        this.companyJobGridDataResult.total === this.selectedCompanyJobIds.length);
  }
}
