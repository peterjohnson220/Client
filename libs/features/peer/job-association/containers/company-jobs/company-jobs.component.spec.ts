import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { of } from 'rxjs';
import { DataStateChangeEvent } from '@progress/kendo-angular-grid';

import { HighlightTextPipe, JobDescriptionParserPipe } from 'libs/core/pipes';
import * as fromRootState from 'libs/state/state';
import * as fromGridActions from 'libs/core/actions/grid.actions';
import { GridTypeEnum } from 'libs/models/common';

import * as fromCompanyJobsActions from '../../actions/company-jobs.actions';
import * as fromCompanyJobsReducer from '../../reducers';
import { CompanyJobsComponent } from './company-jobs.component';

describe('CompanyJobsComponent', () => {
  let component: CompanyJobsComponent;
  let fixture: ComponentFixture<CompanyJobsComponent>;
  let store: Store<fromCompanyJobsReducer.State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          feature_job_association: combineReducers(fromCompanyJobsReducer.reducers)
        }),
      ],
      declarations: [ CompanyJobsComponent, HighlightTextPipe, JobDescriptionParserPipe ],
      schemas: [ NO_ERRORS_SCHEMA ]
    })
      .compileComponents();

    store = TestBed.get(Store);
    spyOn(store, 'dispatch');
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CompanyJobsComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should show total company jobs', () => {
    component.totalCompanyJobsGridItems$ = of(123);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should fire the right actions when the grid state is changed', () => {
    const dataStateChangeEvent = {} as DataStateChangeEvent;
    component.handleDataStateChange(dataStateChangeEvent);

    const updateGridAction = new fromGridActions.UpdateGrid(GridTypeEnum.JobAssociationModalCompanyJobs, dataStateChangeEvent);
    const loadCompanyJobsAction = new fromCompanyJobsActions.LoadCompanyJobs();

    expect(store.dispatch).toHaveBeenCalledWith(updateGridAction);
    expect(store.dispatch).toHaveBeenCalledWith(loadCompanyJobsAction);
  });

  it('should fire the UpdateSearchTerm action when the search term is changed', () => {
    const searchTerm = 'changed';
    component.handleSearchBoxValueChanged(searchTerm);

    const searchTermUpdatedAction = new fromCompanyJobsActions.UpdateSearchTerm(searchTerm);

    expect(store.dispatch).toHaveBeenCalledWith(searchTermUpdatedAction);
  });

  it('should not fire the LoadCompanyJobs action when the search term is one character', () => {
    const searchTerm = 'x';
    component.handleSearchBoxValueChanged(searchTerm);

    const loadCompanyJobsAction = new fromCompanyJobsActions.LoadCompanyJobs();

    expect(store.dispatch).not.toHaveBeenCalledWith(loadCompanyJobsAction);
  });

  it('should fire the updateSearchTermAction followed by loadExchangeJobsAction when the search term is more than one character', () => {
    const searchTerm = 'xx';
    component.handleSearchBoxValueChanged(searchTerm);

    const searchTermUpdatedAction = new fromCompanyJobsActions.UpdateSearchTerm(searchTerm);
    const loadCompanyJobsAction = new fromCompanyJobsActions.LoadCompanyJobs();

    expect(store.dispatch).toHaveBeenCalledWith(searchTermUpdatedAction);
    expect(store.dispatch).toHaveBeenCalledWith(loadCompanyJobsAction);
  });
});
