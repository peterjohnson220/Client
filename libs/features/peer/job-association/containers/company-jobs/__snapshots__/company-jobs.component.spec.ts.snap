// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyJobsComponent should show total company jobs 1`] = `
<pf-peer-job-association-company-jobs
  badRequestError=""
  badRequestError$={[Function Store]}
  badRequestErrorSubscription={[Function Subscriber]}
  companyJobGridDataResult={[Function Object]}
  companyJobSearchComponent="undefined"
  companyJobsGridItemsData$={[Function Store]}
  companyJobsGridItemsDataSubscription={[Function Subscriber]}
  exchangeJobAssociations={[Function Array]}
  exchangeJobAssociationsSubscription={[Function Subscriber]}
  grid="undefined"
  gridState$={[Function Store]}
  isDetailPanelExpanded$={[Function Store]}
  isRowSelected={[Function Function]}
  loading$={[Function Store]}
  loadingError$={[Function Store]}
  maxSelectionThreshold={[Function Number]}
  searchTerm=""
  searchTermSubscription={[Function Subscriber]}
  selectedCompanyJobIds={[Function Array]}
  selectedCompanyJobInDetailPanel={[Function Object]}
  selectedCompanyJobInDetailPanel$={[Function Store]}
  selectedCompanyJobInDetailPanelSubscription={[Function Subscriber]}
  selectedCompanyJobsSubscription={[Function Subscriber]}
  store={[Function Store]}
  tooltipDir="undefined"
  totalCompanyJobsGridItems$={[Function Store]}
>
  <h2
    class="text-center mb-3"
  >
    Company Jobs
  </h2><div
    class="filters position-relative d-flex mt-2 mb-2"
  >
    <i
      class="fas fa-search position-absolute"
    />
    <pf-input-debounce
      id="company-jobs-search-box"
    />
  </div><div>
    <label
      class="text-uppercase"
      data-qa-id="lbl-company-jobs-total"
    >
       company jobs: 0 total | selected: 0 
    </label>
  </div><div
    class="grid-container position-relative"
  >
    <pf-async-container
      class="h-100"
    >
      <div
        class="k-tooltip-container"
        filter=".k-grid td"
        kendotooltip=""
        showon="none"
      >
        <pf-grid-detail-panel />
        <kendo-grid
          style="height: 100%;"
        >
          <kendo-grid-checkbox-column
            showselectall="true"
          >
            
          </kendo-grid-checkbox-column>
          <kendo-grid-column
            field="CompanyJobId"
            hidden="hidden"
            title="Id"
          />
          <kendo-grid-column
            field="JobTitle"
            title="Job Title"
          >
            
          </kendo-grid-column>
          <kendo-grid-column
            field="JobCode"
            title="Job Code"
          >
            
          </kendo-grid-column>
          <kendo-grid-column
            field="JobFamily"
            title="Job Family"
          >
            
          </kendo-grid-column>
          <kendo-grid-column
            field="AssociationStatus"
            title="Status"
          >
            
          </kendo-grid-column>
          
        </kendo-grid>
      </div>
    </pf-async-container>
  </div>
</pf-peer-job-association-company-jobs>
`;
